prqlc/prqlc/src/cli/highlight.rs

```diff
@@ -0,0 +1,128 @@
+use color_eyre::owo_colors::OwoColorize;
+use prqlc::{
+    lr::{TokenKind, Tokens},
+    pr::Literal,
+};
+
+/// Highlight PRQL code printed to the terminal.
+pub fn highlight(tokens: &Tokens) -> String {
+    let mut output = String::new();
+    let mut last = 0;
+
+    for token in &tokens.0 {
+        let diff = token.span.start - last;
+        last = token.span.end;
+        output.push_str(&" ".repeat(diff));
+
+        match &token.kind {
+            TokenKind::NewLine => output.push('\n'),
+            TokenKind::Ident(ident) => {
+                if is_transform(ident) {
+                    output.push_str(&ident.green().to_string())
+                } else {
+                    output.push_str(&ident.to_string())
+                }
+            }
+            TokenKind::Keyword(keyword) => output.push_str(&keyword.blue().to_string()),
+            TokenKind::Literal(literal) => output.push_str(&match literal {
+                Literal::Null => literal.green().bold().to_string(),
+                Literal::Integer(_) => literal.green().to_string(),
+                Literal::Float(_) => literal.green().to_string(),
+                Literal::Boolean(_) => literal.green().bold().to_string(),
+                Literal::String(_) => literal.yellow().to_string(),
+                _ => literal.to_string(),
+            }),
+            TokenKind::Param(param) => output.push_str(&param.purple().to_string()),
+            TokenKind::Range {
+                bind_left: _,
+                bind_right: _,
+            } => output.push_str(".."),
+            TokenKind::Interpolation(_, _) => output.push_str(&format!("{}", token.kind.yellow())),
+            TokenKind::Control(char) => output.push(*char),
+            TokenKind::ArrowThin
+            | TokenKind::ArrowFat
+            | TokenKind::Eq
+            | TokenKind::Ne
+            | TokenKind::Gte
+            | TokenKind::Lte
+            | TokenKind::RegexSearch => output.push_str(&format!("{}", token.kind)),
+            TokenKind::And | TokenKind::Or => {
+                output.push_str(&format!("{}", token.kind).purple().to_string())
+            }
+            TokenKind::Coalesce | TokenKind::DivInt | TokenKind::Pow | TokenKind::Annotate => {
+                output.push_str(&format!("{}", token.kind))
+            }
+            TokenKind::Comment(comment) => output.push_str(
+                &format!("#{comment}")
+                    .truecolor(95, 135, 135)
+                    .italic()
+                    .to_string(),
+            ),
+            TokenKind::DocComment(comment) => output.push_str(
+                &format!("#!{comment}")
+                    .truecolor(95, 135, 135)
+                    .italic()
+                    .to_string(),
+            ),
+            TokenKind::LineWrap(_) => todo!(),
+            TokenKind::Start => {}
+        }
+    }
+
+    output
+}
+
+fn is_transform(ident: &str) -> bool {
```

Comment:
```suggestion
fn is_transform(ident: &str) -> bool {
// TODO: Could we instead source these from the standard library?
// We could also use the semantic understanding from later compiler stages?
```

Reply:
I don't think we can use the semantic understanding from a later compiler stage because a lexer is dumb and only break downs tokens while a later compiler stage requires the token stream to be valid in order to extract out semantics from it, hence it would only work on a valid PRQL query.

Syntax highlighting is usually dumb and carried out on tokens produced from a lexer or regular expressions because it is fail-safe and fast where as relying on a compiler is slow and only works if the the input is valid.